p2p/p2p.md

```diff
@@ -0,0 +1,222 @@
+
+* with max height among the received
+* which is received from at least `minHeadResponses` of 2 peers
+* when neither or both conditions meet, the head with highest height is used
+
+Apart from requesting the latest header, any arbitrary header(s) can be requested (with 3 retries) using height (`GetByHeight`), hash (`Get`), range (`GetRangeByHeight` and `GetVerifiedRange`) from trusted peers as defined in the request proto message:
+
+```
+message HeaderRequest {
+  oneof data {
+    uint64 origin = 1;
+    bytes hash = 2;
+  }
+  uint64 amount = 3;
+}
+```
+
+The `GetVerifiedRange` differs from `GetRangeByHeight` as it ensures that the returned headers are correct against another header (passed as parameter to the call).
+
+### Exchange Server
+
+ExchangeServer represents the server-side component (p2p server) for responding to inbound header requests. The exchange server needs to be initialized using self [host.Host][host] and a [store](../specs/src/specs/store.md). Optional `ServerParameters` as shown below, can be set during the server initialization.
+
+```
+// ServerParameters is the set of parameters that must be configured for the exchange.
+type ServerParameters struct {
+ // WriteDeadline sets the timeout for sending messages to the stream
+ WriteDeadline time.Duration
+ // ReadDeadline sets the timeout for reading messages from the stream
+ ReadDeadline time.Duration
+ // RangeRequestTimeout defines a timeout after which the session will try to re-request headers
+ // from another peer.
+ RangeRequestTimeout time.Duration
+ // networkID is a network that will be used to create a protocol.ID
+ // Is empty by default
+ networkID string
+}
+```
+
+The default values for `ServerParameters` are as described below.
+
+```
+// DefaultServerParameters returns the default params to configure the store.
+func DefaultServerParameters() ServerParameters {
+ return ServerParameters{
+  WriteDeadline:       time.Second * 8,
+  ReadDeadline:        time.Minute,
+  RangeRequestTimeout: time.Second * 10,
+ }
+}
+```
+
+During the server start, a request handler for the `protocolID` (`/networkID/header-ex/v0.0.3`) which defined using the `networkID` configurable parameter is setup to serve the inbound header requests.
+
+The request handler returns a response which contains bytes of the requested header(s) and a status code as shown below.
+
+```
+message HeaderResponse {
+  bytes body = 1;
+  StatusCode statusCode = 2;
+}
+```
+
+The `OK` status code for success, `NOT_FOUND` for requested headers not found, and `INVALID` for error (default).
+
+```
+enum StatusCode {
+  INVALID = 0;
+  OK = 1;
+  NOT_FOUND = 2;
+}
+```
+
+The request handler utilizes its local [store](../specs/src/specs/store.md) for serving the header requests and only up to `MaxRangeRequestSize` of 512 headers can be requested while requesting headers by range. If the requested range is not available, the range is reset to whatever is available.
```

Comment:
we should use an absolute link here. a relative link to a symlink might be buggy with the rendering.